File: python/packages/var-core/src/var_core/registry.py
"""Registry for step definitions — port of var-core/src/registry.ts."""
from __future__ import annotations

from dataclasses import dataclass
from re import Pattern
from typing import Any, Callable, Optional, Union

from cucumber_expressions.expression import CucumberExpression
from cucumber_expressions.parameter_type import ParameterType
from cucumber_expressions.parameter_type_registry import ParameterTypeRegistry

from var_core.step_role import StepKind

# A step handler: receives the context state plus any matched arguments
StepHandler = Callable[..., Any]

# Accepted regexp forms for define_parameter_type
RegexpInput = Union[str, Pattern, list[Union[str, Pattern]]]


@dataclass(frozen=True)
class StepRegistration:
    expression: str
    expression_source_file: str
    expression_source_line: int
    handler: StepHandler
    compiled: CucumberExpression
    kind: Optional[StepKind] = None


@dataclass(frozen=True)
class Registry:
    steps: tuple[StepRegistration, ...]
    parameter_types: ParameterTypeRegistry


def create_registry() -> Registry:
    """Return an empty registry with a fresh default ParameterTypeRegistry."""
    return Registry(steps=(), parameter_types=ParameterTypeRegistry())


def add_step(
    registry: Registry,
    *,
    expression: str,
    expression_source_file: str,
    expression_source_line: int,
    handler: StepHandler,
    kind: Optional[StepKind] = None,
) -> Registry:
    """Compile *expression* and append it to *registry*.

    Returns a new Registry; the original is unchanged (ParameterTypeRegistry is
    shared by reference — see note in TS source — but never mutated here).

    Raises ValueError on duplicate expressions, mirroring the TS error message.
    """
    duplicate = next((s for s in registry.steps if s.expression == expression), None)
    if duplicate is not None:
        raise ValueError(
            f'duplicate step definition for "{expression}" at '
            f"{duplicate.expression_source_file}:{duplicate.expression_source_line} and "
            f"{expression_source_file}:{expression_source_line}"
        )
    compiled = CucumberExpression(expression, registry.parameter_types)
    reg = StepRegistration(
        expression=expression,
        expression_source_file=expression_source_file,
        expression_source_line=expression_source_line,
        handler=handler,
        compiled=compiled,
        kind=kind,
    )
    return Registry(steps=(*registry.steps, reg), parameter_types=registry.parameter_types)


def define_parameter_type(
    registry: Registry,
    *,
    name: str,
    regexp: RegexpInput,
    transformer: Optional[Callable[..., Any]] = None,
    use_for_snippets: bool = True,
    prefer_for_regexp_match: bool = False,
) -> Registry:
    """Register a custom parameter type with the shared ParameterTypeRegistry.

    The registry's ParameterTypeRegistry is mutated in place (same semantics as
    the TS implementation), so previously compiled CucumberExpressions also gain
    the new type.

    Returns the same Registry object (the mutation is in-place).
    """
    # Normalise regexp to a list so ParameterType always receives a list
    if isinstance(regexp, (str, Pattern)):
        regexps: list[str | Pattern] = [regexp]
    else:
        regexps = list(regexp)

    # Default transformer: identity — return the first (and usually only) group
    effective_transformer: Callable[..., Any] = transformer if transformer is not None else (
        lambda *groups: groups[0]
    )

    pt: ParameterType = ParameterType(
        name,
        regexps,
        None,
        effective_transformer,
        use_for_snippets,
        prefer_for_regexp_match,
    )
    registry.parameter_types.define_parameter_type(pt)
    # Return the same object — the mutation is in the ParameterTypeRegistry
    return registry

File: java/var-core/src/main/java/dev/varar/core/Registry.java
package dev.varar.core;

import io.cucumber.cucumberexpressions.CaptureGroupTransformer;
import io.cucumber.cucumberexpressions.Expression;
import io.cucumber.cucumberexpressions.ExpressionFactory;
import io.cucumber.cucumberexpressions.ParameterType;
import io.cucumber.cucumberexpressions.ParameterTypeRegistry;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.function.Function;
import java.util.regex.Pattern;

/**
 * Step registry — port of {@code var-core/src/registry.ts}. Wraps the Java {@code
 * cucumber-expressions} library ({@code io.cucumber:cucumber-expressions:20.0.0}).
 *
 * <p><b>API-surface note, confirmed against the 20.0.0 jar (via {@code javap -p}), not
 * assumed:</b> the TS/Python ports construct {@code CucumberExpression} from a public
 * constructor. The Java library's {@code CucumberExpression(String,
 * ParameterTypeRegistry)} constructor is package-private — {@code javap} prints it with
 * no access modifier, and the class is otherwise unreachable for direct construction
 * from {@code dev.varar.core}. The library's public entry point is instead {@code
 * ExpressionFactory(ParameterTypeRegistry).createExpression(String)}, which returns the
 * public {@code Expression} interface (dispatching internally to a {@code
 * CucumberExpression} unless the source string is itself an anchored/regex-literal form
 * — {@code ^...$} or {@code /.../} — which this project's step expressions never are;
 * confirmed by reading {@code ExpressionFactory}'s decompiled bytecode). {@link
 * StepRegistration#compiled} is therefore typed as {@code Expression}, not the
 * (inaccessible) concrete {@code CucumberExpression} — {@code Expression}'s {@code
 * match}/{@code getRegexp}/{@code getSource} are exactly what Task 14's matcher needs
 * regardless of the concrete implementation.
 *
 * <p>{@code ParameterTypeRegistry} likewise has no no-arg constructor in this version —
 * only {@code ParameterTypeRegistry(Locale)}. {@link Locale#ENGLISH} is used here; this
 * only affects locale-sensitive built-in types ({@code float}/{@code double} decimal
 * separators), not {@code int}/{@code word}/{@code string}.
 */
public record Registry(
        List<StepRegistration> steps,
        ParameterTypeRegistry parameterTypes,
        List<CustomParameterType> customParameterTypes,
        Map<String, Function<Object, String>> formats) {

    public Registry {
        steps = List.copyOf(steps);
        customParameterTypes = List.copyOf(customParameterTypes);
        formats = Map.copyOf(formats);
    }

    /**
     * A custom parameter type as registered by an author — name plus the bare pattern
     * source ({@link Pattern#pattern()}, no flags/delimiters), the exact string the
     * conformance registry artifact serializes. Built-ins never appear here.
     */
    public record CustomParameterType(String name, String regexp) {}

    /**
     * One registered step: its source cucumber expression, source location, handler
     * (retained type-erased as {@link Object} — execution is a later task), the compiled
     * expression, and its role. {@code kind} may be {@code null} (the legacy/kindless
     * step path, mirroring TS/Python's optional {@code kind}).
     */
    public record StepRegistration(
            String expression,
            String expressionSourceFile,
            int expressionSourceLine,
            Object handler,
            Expression compiled,
            StepKind kind) {}

    /** An empty registry with a fresh default {@link ParameterTypeRegistry}. */
    public static Registry createRegistry() {
        return new Registry(List.of(), new ParameterTypeRegistry(Locale.ENGLISH), List.of(), Map.of());
    }

    /**
     * Compiles {@code expression} against {@code registry}'s parameter types and appends
     * it, returning a new {@link Registry}; the original is unchanged.
     *
     * @throws IllegalArgumentException if {@code expression} duplicates an existing
     *     registration in {@code registry} — the message lists both source positions,
     *     mirroring {@code registry.ts}'s error text.
     */
    public static Registry addStep(
            Registry registry,
            String expression,
            String expressionSourceFile,
            int expressionSourceLine,
            Object handler,
            StepKind kind) {
        for (StepRegistration existing : registry.steps()) {
            if (existing.expression().equals(expression)) {
                throw new IllegalArgumentException("duplicate step definition for \""
                        + expression
                        + "\" at "
                        + existing.expressionSourceFile()
                        + ":"
                        + existing.expressionSourceLine()
                        + " and "
                        + expressionSourceFile
                        + ":"
                        + expressionSourceLine);
            }
        }
        Expression compiled = new ExpressionFactory(registry.parameterTypes()).createExpression(expression);
        List<StepRegistration> next = new ArrayList<>(registry.steps());
        next.add(new StepRegistration(expression, expressionSourceFile, expressionSourceLine, handler, compiled, kind));
        return new Registry(next, registry.parameterTypes(), registry.customParameterTypes(), registry.formats());
    }

    /**
     * Registers a custom parameter type on {@code registry}'s shared {@link
     * ParameterTypeRegistry} and returns a NEW {@link Registry} recording it in {@link
     * #customParameterTypes()}. The underlying {@link ParameterTypeRegistry} is still
     * mutated in place — {@code ParameterTypeRegistry.defineParameterType} mutates in
     * place, and (as in the TS/Python ports) that mutation is intentionally shared across
     * every step subsequently compiled against this registry — but callers MUST use the
     * returned {@link Registry} (not the argument) to observe the tracked custom type.
     *
     * @param parse maps the matched capture group(s) to the argument value; unlike
     *     TS/Python there is no implicit identity default — callers pass one explicitly
     */
    public static <T> Registry defineParameterType(
            Registry registry, String name, Pattern regexp, Function<String[], T> parse) {
        return defineParameterType(registry, name, regexp, parse, null);
    }

    /**
     * As {@link #defineParameterType(Registry, String, Pattern, Function)}, additionally
     * retaining {@code format} — the inverse of {@code parse}, rendering a value back in
     * the document's notation. Display-only: it is consulted solely when a sensor's
     * returned value mismatches a transformed inline parameter (see {@link ParamDiff});
     * it never influences matching or the comparison verdict. Kept in a parallel
     * name-keyed map on this record because the underlying {@code cucumber-expressions}
     * {@link ParameterType} cannot carry it — mirroring the TS port's {@code formats}
     * map beside its {@code ParameterTypeRegistry}.
     *
     * @param parse maps the matched capture group(s) to the argument value
     * @param format renders a value of this type in the document's notation; {@code
     *     null} means no formatter (mismatches fall back to the generic rendering chain)
     */
    public static <T> Registry defineParameterType(
            Registry registry, String name, Pattern regexp, Function<String[], T> parse, Function<T, String> format) {
        CaptureGroupTransformer<T> adapted = groups -> parse.apply(groups);
        // TS passes `null` for `type` (a JS ParameterType tolerates it, deanonymizing
        // later); this library's constructor `Objects.requireNonNull`s it (confirmed via
        // decompiled bytecode) — `Object.class` is the same stand-in the library's own
        // `createAnonymousParameterType` passes when no concrete type is known.
        ParameterType<T> parameterType = new ParameterType<>(
                name,
                List.of(regexp.pattern()),
                (Type) Object.class,
                adapted,
                /* useForSnippets= */ true,
                /* preferForRegexpMatch= */ false,
                /* useRegexpMatchAsStrongTypeHint= */ false);
        registry.parameterTypes().defineParameterType(parameterType);
        List<CustomParameterType> recorded = new ArrayList<>(registry.customParameterTypes());
        recorded.add(new CustomParameterType(name, regexp.pattern()));
        Map<String, Function<Object, String>> formats = registry.formats();
        if (format != null) {
            Map<String, Function<Object, String>> next = new LinkedHashMap<>(registry.formats());
            @SuppressWarnings("unchecked")
            Function<Object, String> erased = (Function<Object, String>) format;
            next.put(name, erased);
            formats = next;
        }
        return new Registry(registry.steps(), registry.parameterTypes(), recorded, formats);
    }
}
